components/hab/src/command/pkg/uninstall.rs

```diff
@@ -217,7 +217,11 @@ fn do_clean_delete(pkg_root_path: &Path, real_install_path: &Path) -> Result<boo
     // below the pkg_root_path
     match real_install_path.parent() {
         Some(real_install_base) => {
+            // create empty temp directory
             let temp_install_path = temp_package_directory(real_install_path)?;
+            // on windows fs::rename cannot rename to an existing directory even if empty
+            #[cfg(windows)]
+            fs::remove_dir(&temp_install_path)?;
```

Comment:
So, what we're really saying here is that we need `temp_package_directory` to provide us with a suitable _path_, rather than an actual directory, correct?

Since `temp_package_directory` returns a [TempDir](https://docs.rs/tempfile/3.0.7/tempfile/struct.TempDir.html), we may be able to take advantage of it's `Drop` behavior here. I'm thinking something like this:

```rust
let temp_install_path = {
    let td = temp_package_directory(real_install_path)?;
    td.into_path()
};
```

or, more succinctly:

```rust
 let temp_install_path = temp_package_directory(real_install_path)?.into_path();
```

That should end up removing the directory, and giving you the path you need, and it should be the same on both platforms.

Reply:
but we only want to do that on windows. We actually do need the directory on linux. In that case I'm wondering if just the single platform constrained delete line is better.

Reply:
Unless I'm missing something, you don't actually need the empty directory on Linux in order to rename it.

```sh
maier@ubuntu:~/testing$ ls
test.rs
maier@ubuntu:~/testing$ cat test.rs 
use std::fs;

fn main() {
    fs::create_dir_all("foo").expect("couldn't create foo");
    fs::rename("foo", "bar").expect("couldn't rename to bar");
}
maier@ubuntu:~/testing$ rustc test.rs 
maier@ubuntu:~/testing$ ls
test  test.rs
maier@ubuntu:~/testing$ ./test
maier@ubuntu:~/testing$ ls
bar  test  test.rs
maier@ubuntu:~/testing$ ls -alh bar
total 8.0K
drwxr-xr-x 2 maier maier 4.0K Mar 20 16:04 .
drwxr-xr-x 3 maier maier 4.0K Mar 20 16:04 ..
maier@ubuntu:~/testing$ touch bar/blah
maier@ubuntu:~/testing$ ./test
thread 'main' panicked at 'couldn't rename to bar: Os { code: 39, kind: Other, message: "Directory not empty" }', src/libcore/result.rs:997:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
maier@ubuntu:~/testing$ 

```

I think the docs are saying that if the `to` directory exists on Linux, it has to be empty.

Reply:
oh ok. cool. I will try that!

Reply:
Ended up that `into_path` "persists" the `TempDir` struct in which case its `drop` does not remove the file. However `path().to_path_buf()` accomplishes the same thing and does allow the `drop` to remove the directory. I validated this on both windows and linux.

Reply:
Oh, TIL 😄